man/lib/c/Mig.man
\fB'\" $Header: /sprite/src/lib/c/mig/RCS/Mig.man,v 1.2 90/02/28 11:00:30 douglis Exp $ SPRITE (Berkeley)
.so \*(]ltmac.sprite
.HS Mig lib
.BS
.SH NAME
Mig \- obtain or update information in the migration information database.
.SH SYNOPSIS
.nf
\fB#include	<mig.h>\fR

Mig_Info *
\fBMig_GetInfo\fR(\fIspriteID\fP)

int
\fBMig_GetAllInfo\fR(\fIinfoArray, numHosts\fP)

int
\fBMig_RequestIdleHosts\fR(\fInumHosts, priority, flags, callBackPtr, hostArray\fP)

int
\fBMig_ConfirmIdle\fR(\fIspriteID\fP)

int
\fBMig_ReturnHosts\fR(\fInumHosts, hostArray\fP)

int
\fBMig_DeleteHost\fR(\fIspriteID\fP)

int
\fBMig_Evict\fR()

char *
\fBMig_GetPdevName\fR(\fIglobal\fP)

int
\fBMig_GetIdleNode\fR()

int
\fBMig_Done\fR(\fIspriteID\fP)

.SH ARGUMENTS
.AS void (*callBackPtr)() in
.AP int spriteID in
The Sprite ID of the host for which information should be obtained or updated.
.AP Mig_Info infoArray[] out
A buffer to hold Mig_Info entries returned by the Mig_GetAllInfo routine.
The size of the buffer is specified in the \fInumHosts\fP variable.
.AP int numHosts in
The number of Mig_Info structures contained in \fIinfoArray\fP.
Or, the number of hosts requested from \fBMig_RequestIdleHosts\fR and
the number of hosts returned to \fBMig_ReturnIdle\fR.  
.AP int priority in
Priority of processes to be migrated to idle hosts (see below).  
.AP int flags in
Flags to be passed to the migration daemon (see below).
.AP void (*callBackPtr)() in
Routine to call if not enough hosts are available and
additional hosts are later available.
.AP int global in
Whether to return the name of the pseudo-device for the global daemon
or host-specific daemon.
.BE
.SH DESCRIPTION
These functions are used to obtain (or update) information about hosts on the 
Sprite network.  There are routines to get information about specific
hosts or all the hosts on the network, to select one or more idle hosts for
process migration, to cause processes to be evicted, or to remove
hosts that are down from the list of hosts on the network.  
.PP
Most of the routines interact with a single network-wide daemon, known
as the \fIglobal daemon\fP, to obtain information or make requests.
This \fBmigd\fP daemon maintains state about all hosts on the network,
including their load averages, idle time, and availability for use
with migration.   In addition, each host runs a per-host instance of
\fBmigd\fR that samples load and evicts processes automatically when
it detects user input.    The \fBMig\fR library opens a pseudo-device
to communicate with the appropriate \fBmigd\fR daemon (global or
local) depending on the operation being performed.  Also, if an error
occurs communicating with a daemon, the \fBMig\fR library
reestablishes communication.
.SH GETTING HOST INFORMATION
The migration daemon, and the migration library, communicate using a
structure defined in mig.h, known  as a Mig_Info structure.  
Each host has a Mig_Info structure associated with it.  The structure
has another structure included within it; the Mig_LoadVector is
updated periodically by the \fBmigd\fR process on each host.  The rest
of the data in a Mig_Info structure are established at boot-time or are
maintained by the global daemon.  
.PP
The global daemon maintains the state of each host.  The states are
defined in mig.h.  The most important ones are MIG_HOST_DOWN, which
indicates that the rest of the Mig_Info structure is irrelevant with
the exception of loadVec.timestamp, which indicates when the host was
last up; MIG_HOST_IDLE, indicating that a host is available for
migration; MIG_HOST_FULL, which shows that the host is accepting
migration but already has foreign processes; and MIG_HOST_ACTIVE,
which indicates that a user is actively using the machine or the
machine's load is too high to permit foreign processes.
.PP
The load vector includes weighted
average CPU utilizations and ready-queue lengths, as well as the time since
input was last received from users directly logged into the machine.  (Remote
logins do not affect idle time.)  Finally, each entry indicates whether the
host is willing to accept processes for migration.   \fBmigd\fR 
is responsible for determining whether migration
is allowed, based on factors such as load and idle time.  
.PP
\fBMig_GetInfo()\fR returns a pointer to a Mig_Info structure based on the sprite 
ID of a host. 
The structure is statically allocated by Mig_GetInfo, and the contents of
the structure may change on subsequent calls to Mig_GetInfo.  On
error, a NULL pointer is returned.
.PP
\fBMig_GetAllInfo()\fR returns an array of Mig_Info structures.  The array
must be allocated by the caller, and a pointer to the array must be
passed to \fBMig_GetAllInfo()\fR, along with the size of the array.  The
number of entries filled in \fiinfoArray\fP is returned.  On error, -1
is returned and the global variable \fIerrno\fP indicates the nature
of the error.  
.PP
.SH GETTING IDLE HOSTS
The global daemon maintains open connections to both per-host
\fBmigd\fR daemons and user processes that are using idle hosts.  User
processes can request idle hosts using any of three priorities,
MIG_LOW_PRIORITY, MIG_NORMAL_PRIORITY, and MIG_HIGH_PRIORITY.
MIG_LOW_PRIORITY is for long-running background processes that are to
be run at low execution priority.  MIG_NORMAL_PRIORITY is for normal,
relatively short-lived processes such as compiles.  MIG_HIGH_PRIORITY
is not yet in general use.  Hosts with foreign processes of one
priority can still accept more processes of higher priority, so
simulations won't interfere with compiles.
.PP
Processes communicate with the global daemon using ioctls, and
normally, those streams are
not readable. 
The global daemon makes the streams readable to indicate that a change
of state has occurred.  The \fBMig\fR library 
relies on \fBselect()\fR to determine whether the global daemon has
information about changes.  In this way, communication with the daemon
may be minimized, since applications can check for new idle hosts or
evictions on hosts they were using, without ever communicating with
the daemon. 
.PP
\fBMig_RequestIdleHosts()\fR requests one or more idle hosts from the
global daemon.   The number of hosts available is returned.  The
\fIpriority\fR must be specified as one of the priorities listed
above.  The \fIflags\fR may be 0 or MIG_PROC_AGENT, which indicates
that the process requesting the host will not indicate when the
processes being migrated are through, and instead the hosts to which
the processes are migrated should monitor foreign processes and note
when they are no longer in use for migration.   The \fIcallBackPtr\fR
may be NULL or a pointer to a void function that will be invoked when
additional hosts are available, if an insufficient number of hosts are
granted by the global daemon at the time of the call.  \fIhostArray\fR
points to an area that can hold up to \fInumHosts\fR host identifiers.
.PP
\fBMig_ConfirmIdle(\fIhostID\fB)\fR verifies that a host is still
available.  If 
the host is available, 
\fBMig_ConfirmIdle()\fR returns 1 (TRUE).  If the host is not
available, or an error occurs, \fBMig_ConfirmIdle()\fR returns 0 (FALSE).
In this case, the caller may request a new idle host.  
.PP
\fBMig_ReturnHosts()\fR returns one or more idle hosts to the pool.
Note: all hosts requested by a process are returned to the pool of
idle hosts when the stream that connects the process to the global
daemon is closed (i.e., when the process and all its children that may
have inherited the stream have exited).  
.SH BACKWARD COMPATIBILITY
Two functions are implemented to provide a backward compatible
interface for users of the original \fBMig\fR library.  
.PP
\fBMig_GetIdleNode()\fR returns the number of an idle node.  If no host is
available, then 0 is returned.  On error, -1
is returned and the global variable \fIerrno\fP indicates the nature
of the error.  
.PP
\fBMig_Done()\fR returns a single host to the pool of idle hosts.
.SH MISCELLANEOUS
.PP
\fBMig_DeleteHost()\fR removes a host from the database maintained by
the global daemon.  The host must be down at the time.  This may be used
if a host is removed from the network or renamed.
.PP
\fBMig_Evict()\fR performs an \fIioctl\fP to the local \fBmigd\fR
daemon, requesting that it evict any foreign processes.  Normally,
eviction is automatic when a host becomes active after being idle.
This routine provides the \fBloadavg\fP program with the ability to
request evictions at other times (for example, from a remote login).
.PP
\fBMig_GetPdevName()\fR is used by the migration library and by
\fBmigd\fR to get the name of pseudo-devices to open.  It would not
normally be used by other programs.
.SH DIAGNOSTICS
Most \fBMig\fR routines
return zero if 
all went well.  Otherwise 
they return -1 and the \fIerrno\fR variable contains additional information
about what error occurred.  \fBMig_GetIdleNode()\fR  and
\fBMig_RequestIdleHosts()\fR   similarly return -1 on 
error, but they return 0 if no idle host is available.  \fBMig_GetInfo()\fR
returns
NULL on error. 
.SH FILES
.IP /sprite/admin/migd.log
The global migration daemon error log.  
.IP /hosts/$HOST/migd.log
The error log used by host $HOST.
.IP /sprite/admin/migInfo.pdev
The pseudo-device used to communicate with the global daemon.
.IP /hosts/$HOST/migInfo.pdev
The pseudo-device used to communicate with the local daemon.
.IP /sprite/admin/migd.check
The file used to store the most recent information about host uptimes.
.SH KEYWORDS
migration, load average, idle time, pseudo-device, eviction
.SH SEE ALSO
migd, loadavg, pmake, pdev


src/cmds/migrate/mig.c
/* 
 * mig.c --
 *
 *	Program to perform remote execution using process migration.
 *
 * Copyright 1988 Regents of the University of California
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies.  The University of California
 * makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#ifndef lint
static char rcsid[] = "$Header: /sprite/src/cmds/migrate/RCS/mig.c,v 1.14 90/09/24 14:38:37 douglis Exp $ SPRITE (Berkeley)";
#endif not lint


#include <sprite.h>
#include <status.h>
#include <option.h>
#include <proc.h>
#include <stdio.h>
#include <stdlib.h>
#include <host.h>
#include <sys/wait.h>
#include <signal.h>
#include <string.h>
#include <errno.h>
#include <mig.h>

/*
 * Library imports:
 */

extern char **environ;

/*
 * Forward declarations
 */
static void DoExec();
static int RemotePathExec();

int debug = 0;
char *host = NULL;
int hostID = -1;
char *procIDString = NULL;
int local = 0;
int background = 0;
int backPrio = 0;
int verbose = 0;
char *execName = NULL;

Option optionArray[] = {
    {OPT_STRING, "h", (Address) &host,
	 "String identifier for host to migrate onto (default is random selection)."},
    {OPT_INT, "H", (Address) &hostID,
	 "Numeric identifier for host to migrate onto."},
    {OPT_TRUE, "b", (Address) &background,
	 "Execute command in background and print processID."},
    {OPT_TRUE, "B", (Address) &backPrio,
	 "Execute command at background priority."},
    {OPT_TRUE, "l", (Address) &local,
	 "Execute command locally if no idle node is available."},
    {OPT_TRUE, "D", (Address) &debug,
	 "Enable debugging messages."},
    {OPT_STRING, "E", (Address) &execName,
	 "Exec the command specified by this argument, setting argv[0] to the first argument following the option list."},
    {OPT_TRUE, "v", (Address) &verbose,
	 "Print useful information, such as which host is selected."},
    {OPT_STRING, "p", (Address) &procIDString,
	 "Process ID of process to migrate (default is to start shell, or issue command given as remaining arguments)"},
};

/*
 * Default shell to invoke if none found in environment.
 */
#define SHELL "csh"

main(argc, argv)
    int  argc;
    char *argv[];
{
    ReturnStatus status;
    char *myName;
    char **argArray;
    char *shell;
    int pid;
    Host_Entry *hostPtr;
    int i;
    int selectedHost;
    int doCmd;
    int procID;
    int flags;
    
    argc = Opt_Parse(argc, argv, optionArray, Opt_Number(optionArray), 
		     OPT_ALLOW_CLUSTERING|OPT_OPTIONS_FIRST);

    myName = argv[0];
    if (hostID != -1 && host != (char *) NULL) {
	fprintf(stderr, "%s: -h and -H options are mutually exclusive\n",
		myName);
	exit(1);
    }
    if (debug) {
	verbose = 1;
    }
    if (procIDString != (char *) NULL) {
	if (background) {
	    fprintf(stderr, "%s: -b and -p options are mutually exclusive\n",
		    myName);
	    exit(1);
	}
	procID = strtol(procIDString, (char **) NULL, 16);
	if (procID <= 0) {
	    (void)fprintf(stderr, "%s: invalid process ID: %s.\n", myName,
			  procIDString);
	    exit(1);
	}
	doCmd = 0;
	if (argc > 1) {
	    (void)fprintf(stderr, "%s: extra arguments ignored.\n", myName);
	}
	if (local) {
	    (void)fprintf(stderr,
			  "%s: -l argument ignored when migrating process.\n",
			  myName);
	    local = 0;
	}
	flags = MIG_PROC_AGENT;
    } else {
	doCmd = 1;
	procID = PROC_MY_PID;
	if (argc == 1) {
	    argArray = (char **) malloc(2 * sizeof(char *));
	    shell = getenv("SHELL");
	    if (shell == (char *) NULL) {
		shell = SHELL;
	    }
	    argArray[0] = shell;
    	    execName = shell;
	    argArray[1] = NULL;
	} else {
	    argArray = &argv[1];
	    if (!execName) {
		execName = argArray[0];
	    }
	}
	flags = background ? MIG_PROC_AGENT : 0;
    }  
    if (host != (char *) NULL) {
	selectedHost = 0;
	hostPtr = Host_ByName(host);
	if (hostPtr == (Host_Entry *) NULL) {
	    fprintf(stderr, "%s: %s: no such host.\n", myName, host);
	    exit(1);
	}
	hostID = hostPtr->id;
	Host_End();
    } else if (hostID != -1) {
	selectedHost = 0;
    } else {
	int hostNumbers[1];
	int hostsAssigned;
	
	selectedHost = 1;
	(void) signal(SIGINT, SIG_IGN);
	hostsAssigned = Mig_RequestIdleHosts(1, backPrio ?
					     MIG_LOW_PRIORITY :
					     MIG_NORMAL_PRIORITY,
					     flags,
					     (void (*)()) NULL, hostNumbers);
	if (hostsAssigned < 0) {
	    perror("Mig_RequestIdleHosts");
	    exit(1);
	}
	if (hostsAssigned == 0) {
	    if (!local) {
		fprintf(stderr, "%s: no idle host available.\n", myName);
		exit(1);
	    }
	    hostID = 0;
	} else {
	    hostID = hostNumbers[0];
	}
	if (verbose) {
	    if (hostID) {
		hostPtr = Host_ByID(hostID);
		if (hostPtr == (Host_Entry *) NULL) {
		    fprintf(stderr, "%s: unable to find host %d.\n",
			    myName, hostID);
		    exit(1);
		}
		fprintf(stderr, "%s: migrating to host %s.\n", myName,
			hostPtr->name);
		Host_End();
	    } else {
		fprintf(stderr,
			"%s: no idle host available; running locally.\n",
			myName);
	    }
	}
    }

    if ((selectedHost || background) && hostID && doCmd) {
	/*
	 * Wait for the process to complete so we can register that we're
	 * through with the host.
	 */
	pid = fork();
	if (pid < 0) {
	    perror("Error forking child");
	    exit(1);
	}
		
	if (pid) {
	    union wait status;
	    int error;
	    int exited = 0;
	    int exitCode;


	    if (background) {
		printf("%x\n", pid);
		exit(0);
	    }
	    do {
	    
		error = wait(&status);
		if (error == -1) {
		    perror("wait");
		    exit(1);
		} else if (error != pid && debug) {
		    (void) fprintf(stderr, "%s: received status %d from wait.\n",
				   myName, error);
		} else if (error == pid) {
		    if (status.w_stopval != WSTOPPED) {
			exited = 1;
			if (status.w_termsig != 0) {
			    if (debug) {
				(void) fprintf(stderr,
					       "%s: process was signalled.\n",
					       myName);
			    }
			    exitCode = status.w_termsig;
			} else {
			    exitCode = status.w_retcode;
			}
		    } else {
			if (debug) {
			    (void) fprintf(stderr,
					   "%s: process was suspended.\n",
					   myName);
			}
		    }
		}
	    } while (!exited);
#ifdef undef
	    if (debug) {
		(void) fprintf(stderr,
			       "%s: Exited(%d).  Calling Mig_Done(%d).\n",
			       myName, exitCode, hostID);
		(void) fflush(stderr);
	    }
	    (void) Mig_Done(hostID);
#endif
	    exit(exitCode);
	} else {
	    (void) signal(SIGINT, SIG_DFL);
	}
    }
    if (hostID && !doCmd) {
	/*
	 * We're migrating someone else.
	 */
	
	if (debug) {
	    (void) fprintf(stderr, "Calling Proc_Migrate(%d, %d).\n", procID,
			   hostID);
	    (void) fflush(stderr);
	}
	
	status = Proc_Migrate(procID, hostID);
	if (status != SUCCESS) {
	    fprintf(stderr, "%s: error in Proc_Migrate: %s\n",
		    myName, Stat_GetMsg(status));
	    fflush(stderr);
	    if (!local) {
		exit(1);
	    }
	}
    }
    if (!doCmd) {
	exit(0);
    }
    /*
     * Still the child here, or the only process if we were running locally
     * and never forked.
     */
    if (debug) {
	(void) fprintf(stderr, "Calling RemotePathExec(%s,...,%d).\n",
		       execName, hostID);
	(void) fflush(stderr);
    }
    status = RemotePathExec(execName, argArray, hostID);
    perror("Error execing program");
    exit(status);
}


/*
 *-----------------------------------------------------------------------
 *
 * DoExec --
 *
 *	Function to actually execute a program. If the exec didn't succeed
 *	because the file isn't in a.out format, attempt to execute
 *	it as a bourne shell script.
 *
 * Results:
 *	None.  Doesn't even return unless the exec failed.
 *
 * Side Effects:
 *	A program may be execed over this one.
 *
 *-----------------------------------------------------------------------
 */

static void
DoExec(file, argv, hostID)
    char *file;			/* File to execute. */
    char **argv;		/* Arguments to the program. */
    int hostID;			/* ID of host on which to exec */
{
    ReturnStatus status;
    status = Proc_RemoteExec(file, argv, environ, hostID);
    if (debug) {
	fprintf(stderr, "Proc_RemoteExec(\"%s\"): %s\n", file,
		Stat_GetMsg(status));
    }
    errno = Compat_MapCode(status);
    if (errno == ENOEXEC) {
	/*
	 * Attempt to execute the file as a shell script using
	 * the Bourne shell)
	 */
	register char **newargv;
	register int i;

	for (i = 0; argv[i] != 0; i++) {
	    /* Empty loop body */
	}
	newargv = (char **) malloc((unsigned) ((i+1)*sizeof (char *)));
	newargv[0] = "sh";
	newargv[1] = file;
	for (i = 1; argv[i] != 0; i++) {
	    newargv[i+1] = argv[i];
	}
	newargv[i+1] = 0;
	status = Proc_RemoteExec("/sprite/cmds/sh", newargv, environ, hostID);
	errno = Compat_MapCode(status);
    }
}

/*
 *----------------------------------------------------------------------
 *
 * RemotePathExec --
 *
 *	Execute a process, using the current environment variable,
 *	instead of an explicitly-supplied one.  Also, imitate the
 *	shell's actions in trying each directory in a search path
 *	(given by the "PATH" environment variable).  Also, specify
 *	a remote host.  This is taken from the library execve call.
 *
 * Results:
 *	This procedure returns only if the exec fails.  In this case
 *	the return value is -1.
 *
 * Side effects:
 *	Overlays the current process with a new image.  See the man
 *	page for details.
 *
 *----------------------------------------------------------------------
 */

static int
RemotePathExec(name, argv, hostID)
    char *name;			/* Name of file containing program to exec. */
    char **argv;		/* Array of arguments to pass to program. */
    int hostID;			/* ID of host on which to exec */
{
    char *path;
    char *fullName;
    register char *first, *last;
    int size, noAccess;

    noAccess = 0;

    if (index(name, '/') != 0) {
	/*
	 * If the name specifies a path, don't search for it on the search path,
	 * just try and execute it.
	 */
	DoExec(name, argv, hostID);
	return -1;
    }

    path = getenv("PATH");
    if (path == 0) {
	path = "/sprite/cmds";
    }
    fullName = malloc((unsigned) (strlen(name) + strlen(path)) + 2);
    for (first = path; ; first = last+1) {

	/*
	 * Generate the next file name to try.
	 */

	for (last = first; (*last != 0) && (*last != ':'); last++) {
	    /* Empty loop body. */
	}
	size = last-first;
	(void) strncpy(fullName, first, size);
	if (last[-1] != '/') {
	    fullName[size] = '/';
	    size++;
	}
	(void) strcpy(fullName + size, name);

	if (debug) {
	    fprintf(stderr, "Trying DoExec(\"%s\")....\n", fullName);
	}
	DoExec(fullName, argv, hostID);
	if (debug) {
	    fprintf(stderr, "DoExec(\"%s\") => %d.\n", fullName, errno);
	}
	if (errno == EACCES) {
	    noAccess = 1;
	} else if (errno != ENOENT) {
	    break;
	}
	if (*last == 0) {
	    /*
	     * Hit the end of the path. We're done.
	     * If there existed a file by the right name along the search path,
	     * but its permissions were wrong, return FS_NO_ACCESS. Else return
	     * whatever we just got back.
	     */
	    if (noAccess) {
		errno = EACCES;
	    }
	    break;
	}
    }
    free((char *) fullName);
    return -1;
}
